rb/lib/selenium/webdriver/common/interactions/pointer_input.rb

```diff
@@ -0,0 +1,146 @@
+# distributed with this work for additional information
+# regarding copyright ownership.  The SFC licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+require 'securerandom'
+
+module Selenium
+  module WebDriver
+    module Interactions
+      class PointerInput < InputDevice
+        KIND = {mouse: :mouse, pen: :pen, touch: :touch}.freeze
+
+        attr_reader :kind
+        attr_accessor :primary
+
+        def initialize(kind, name: nil, primary: false)
+          super(name)
+          @kind = assert_kind(kind)
+          @primary = primary
+        end
+
+        def type
+          Interactions::POINTER
+        end
+
+        def encode
+          return nil if no_actions?
+          output = {type: type, id: name, actions: @actions.map(&:encode)}
+          params = {primary: primary, pointerType: kind}
+          output[:parameters] = params
+          output
+        end
+
+        def assert_kind(pointer)
+          raise TypeError, "#{pointer.inspect} is not a valid pointer type" unless KIND.key? pointer
+          KIND[pointer]
+        end
+
+        def create_pointer_move(duration: 0, x: 0, y: 0, element: nil, origin: nil)
+          add_action(PointerMove.new(self, duration, x, y, element, origin))
+        end
+
+        def create_pointer_down(button)
+          add_action(PointerPress.new(self, :down, button))
+        end
+
+        def create_pointer_up(button)
+          add_action(PointerPress.new(self, :up, button))
+        end
+
+        def create_pause(duration = nil)
+          add_action(Pause.new(self, duration))
+        end
+
+        def create_pointer_cancel
+          add_action(PointerCancel.new(self))
+        end
+      end
+
+      class PointerPress < Interaction
+        BUTTONS = {left: 0, middle: 1, right: 2}.freeze
+        DIRECTIONS = {down: :pointerDown, up: :pointerUp}.freeze
+
+        def initialize(source, direction, button)
+          super(source)
+          @direction = assert_direction(direction)
+          @button = assert_button(button)
+        end
+
+        def type
+          @direction
+        end
+
+        def assert_button(button)
+          if button.is_a? Symbol
+            raise ArgumentError, "#{button.inspect} is not a valid button!" unless BUTTONS.key? button
+            button = BUTTONS[button]
+          end
+          raise ArgumentError, 'Button number cannot be negative!' unless button >= 0
```

Comment:
Ah yep so that hash is just so that users have an easier way of specifying a standard mouse button without specifying an integer itself (see the `button.is_a?` conditional above that). Buttons specified as integers are also accepted